### ISSUE TITLE
Tekton controller queue processing time

### ISSUE THREAD

**Issue by yhil** [tektoncd/pipeline#4847]
Hello,

I am scheduling taskruns in a namespace with a resource quota which allow only a certain number of taskruns to schedule their pod at the same time. Each taskrun takes a few minutes to complete. 

The taskruns not running are in the queue waiting for free resources for their pod to be scheduled. The following events are reported:

```
  Normal   ExceededResourceQuota  8m25s (x8 over 8m26s)  TaskRun  TaskRun Pod exceeded available resources: pods "exec-tc-21.6-27832-pod" is forbidden: exceeded quota: tekton, requested: requests.memory=3500Mi, used: requests.memory=213700Mi, limited: requests.memory=211Gi
  Warning  InternalError          8m25s (x8 over 8m26s)  TaskRun  1 error occurred:
           * pods "exec-tc-21.6-27832-pod" is forbidden: exceeded quota: tekton, requested: requests.memory=3500Mi, used: requests.memory=213700Mi, limited: requests.memory=211Gi
```

# Expected Behavior
Tekton controller continuously check if he can process taskrun in the queue. It starts taskrun as soon as resourcequota is free. The number of taskruns running at the same time should be the consistent over time.

# Actual Behavior
Taskrun are waiting in the queue even if resourcequota is free to schedule their pods. The number of running taskrun decrease over time.

# Steps to Reproduce the Problem

1. Create a resource quota in the namespace
2. Create a task with a step using resources requests and limits
3. Apply taskruns which trigerred the previous task. The sum of resources used by the created taskrun should exceed the quota.

# Additional Info

- Kubernetes version:
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v0.21.0-beta.1", GitCommit:"88e7eba36ac7cc026d064bb6e4a988acc18a144e", GitTreeState:"clean", BuildDate:"2022-01-21T03:47:14Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.2-1555+9e8f924492b7d7-dirty", GitCommit:"9e8f924492b7d70b0216d20529f3471eed59463c", GitTreeState:"dirty", BuildDate:"2021-11-10T16:00:46Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
```

- Tekton Pipeline version:
```
Client version: 0.21.0
Pipeline version: v0.34.0
Dashboard version: v0.24.1
```

**Comment by imjasonh** [comment]
We've heard this issue before, and digging into it I believe the underlying issue is that Tekton's retry logic when Pods don't schedule due to resource quota is to retry with exponential backoff.

So, imagine the scenario where a TaskRun is about to be started, but there aren't enough resources right now, but there will be resources available in a few minutes. The TaskRun fails to create, and retries in 1 second. Still no resources, so backoff and retry in 2 seconds. Again, in 4s, then 8s, 16s, 32s, 1 minute, 2 minutes, 4 minutes, 8 minutes (this is an example). If resources became available right now, the TaskRun in the queue wouldn't check to find out for another 8 minutes (or more, if it had retried more times). When it retries, it may find available resources and try again, but if not, it'll try again in 16 minutes, and so on.

This is mostly conjecture, but I think it's a reasonable guess since we do exponential backoff for other failure scenarios.

If this is indeed the issue, we should see if we can configure our backoff logic to cap out at, say, 1 minute, so it never ends up waiting so long to try again.

**Comment by yhil** [comment]
Yes this is what you explained me on Slack, I couldn't find by myself what cause the exponential backoff. Then I created this issue with more detailed to get help and find the root cause of that problem. 

Is it possible to test and confirm your conjecture ?

**Comment by yhil** [comment]
I reapplied the taskruns with the controller log level set to debug. I couldn't see any logs reporting `Failed to create taskrun pod for taskrun .*`
The logs that I see about the ExceededResourceQuota are
```
{
  "level": "debug",
  "ts": "2022-05-11T15:33:39.250Z",
  "logger": "tekton-pipelines-controller",
  "caller": "taskrun/reconciler.go:323",
  "msg": "Updating status with:   v1beta1.TaskRunStatus{\n  \tStatus: v1beta1.Status{\n  \t\tObservedGeneration: 0,\n  \t\tConditions: v1beta1.Conditions{\n  \t\t\t{\n  \t\t\t\tType:               \"Succeeded\",\n  \t\t\t\tStatus:             \"Unknown\",\n  \t\t\t\tSeverity:           \"\",\n- \t\t\t\tLastTransitionTime: apis.VolatileTime{Inner: v1.Time{Time: s\"2022-05-11 15:30:57 +0000 UTC\"}},\n+ \t\t\t\tLastTransitionTime: apis.VolatileTime{\n+ \t\t\t\t\tInner: v1.Time{Time: s\"2022-05-11 15:33:39.249508293 +0000 UTC m=+9104.073853021\"},\n+ \t\t\t\t},\n- \t\t\t\tReason:  \"ExceededResourceQuota\",\n+ \t\t\t\tReason:  \"Pending\",\n- \t\t\t\tMessage: `TaskRun Pod exceeded available resources: pods \"exec-tc-21.6-22901-pod\" is forbidden: exceeded quota: tekton, requested: requests.memory=3500Mi, used: requests.memory=213500Mi, limited: requests.memory=211Gi`,\n+ \t\t\t\tMessage: \"Pending\",\n  \t\t\t},\n  \t\t},\n  \t\tAnnotations: nil,\n  \t},\n  \tTaskRunStatusFields: v1beta1.TaskRunStatusFields{\n- \t\tPodName:         \"\",\n+ \t\tPodName:         \"exec-tc-21.6-22901-pod\",\n- \t\tStartTime:       nil,\n+ \t\tStartTime:       s\"2022-05-11 15:33:39.236575377 +0000 UTC m=+9104.060920114\",\n  \t\tCompletionTime:  nil,\n- \t\tSteps:           nil,\n+ \t\tSteps:           []v1beta1.StepState{},\n  \t\tCloudEvents:     nil,\n  \t\tRetriesStatus:   nil,\n  \t\tResourcesResult: nil,\n  \t\tTaskRunResults:  nil,\n- \t\tSidecars:        nil,\n+ \t\tSidecars:        []v1beta1.SidecarState{},\n  \t\tTaskSpec:        &{Params: {{Name: \"application_version\", Type: \"string\"}, {Name: \"amos_version\", Type: \"string\"}, {Name: \"test_project\", Type: \"string\"}, {Name: \"test_plan\", Type: \"string\"}, ...}, Steps: {{Container: {Name: \"qftest\", Image: \"image-registry.openshift-image-registry.svc:5000/$(params.regist\"..., Env: {{Name: \"AMOS_VERSION\", Value: \"$(params.amos_version)\"}, {Name: \"APPLICATION_VERSION\", Value: \"$(params.application_version)\"}, {Name: \"TEST_PROJECT\", Value: \"$(params.test_project)\"}, {Name: \"TEST_PLAN\", Value: \"$(params.test_plan)\"}, ...}, Resources: {Limits: {s\"memory\": {i: {...}, s: \"3700Mi\", Format: \"BinarySI\"}}, Requests: {s\"cpu\": {i: {...}, s: \"800m\", Format: \"DecimalSI\"}, s\"memory\": {i: {...}, s: \"3500Mi\", Format: \"BinarySI\"}}}, ...}}}},\n  \t},\n  }\n",
  "commit": "e055233",
  "knative.dev/controller": "github.com.tektoncd.pipeline.pkg.reconciler.taskrun.Reconciler",
  "knative.dev/kind": "tekton.dev.TaskRun",
  "knative.dev/traceid": "5240389b-fbe3-4251-afd4-280d588cc9d6",
  "knative.dev/key": "tekton-run-full-single/exec-tc-21.6-22901"
}
```
I also receive logs about requeing, I don't understand them because the time reported is over 50 minutes but the taskrun where in the queue for less than 20 minutes:
```
{
  "level": "debug",
  "ts": "2022-05-11T15:36:54.469Z",
  "logger": "tekton-pipelines-controller",
  "caller": "controller/controller.go:562",
  "msg": "Requeuing key tekton-run-full-single/exec-tc-21.6-19485 (by request) after 57m29.542598207s (depth: 0)",
  "commit": "e055233",
  "knative.dev/controller": "github.com.tektoncd.pipeline.pkg.reconciler.taskrun.Reconciler",
  "knative.dev/kind": "tekton.dev.TaskRun",
  "knative.dev/traceid": "f241ae99-6dab-4323-afa4-dee7fcbc58ce",
  "knative.dev/key": "tekton-run-full-single/exec-tc-21.6-19485"
}
```

**Comment by imjasonh** [comment]
That 57m requeue might be due to this code:

https://github.com/tektoncd/pipeline/blob/155179bf2de895b0ab76859b3d8aef142f0206db/pkg/reconciler/taskrun/taskrun.go#L179-L184

This enqueues a reconcile at a time after the TaskRun should have either been completed or timed out -- if it's not done at that time, it's failed as timed-out.

That re-enqueueing happens only if the TaskRun has a start time, and this code is supposed to unset `.Status.StartTime` if pod creation failed due to resource quotas.

**Comment by yhil** [comment]
I tested the change using two different Pipelines, one created 200 TaskRuns and the other ~ 6'000 in the queue. In my cluster only 65 pods could be scheduled at the same time, leaving all the others TaskRuns with .status.message:
```
TaskRun Pod exceeded available resources: pods "exec-tc-xxxx-pod" is forbidden: exceeded quota: tekton, requested: requests.memory=3500Mi, used: requests.memory=213700Mi, limited: requests.memory=211Gi
```

It resulted in a consistent number of Pods running at the same time until all TaskRun in the queue are proceeded.

I observed only a small raise in term of cpu or memory consumption for the controller pods for 6'000 TaksRuns in the queue: <0.300 cpu and < 700MiB ram.

For the test with thousands of TaskRuns in the queue, I saw that the TaskRuns took more than a minute to be completed with .status.conditions.reason|type: Succeeded. The controller seems to operate all the requeuing before moving the another action.
